src/predictor/gpu_predictor.cu

```diff
@@ -447,6 +456,61 @@ class GPUPredictor : public xgboost::Predictor {
     }
   }
 
+  void PredictContribution(DMatrix* p_fmat,
+                           std::vector<bst_float>* out_contribs,
+                           const gbm::GBTreeModel& model, unsigned ntree_limit,
+                           std::vector<bst_float>* tree_weights,
+                           bool approximate, int condition,
+                           unsigned condition_feature) override {
+    if (approximate) {
+      LOG(FATAL) << "[Internal error]: " << __func__
+                 << " approximate is not implemented in GPU Predictor.";
+    }
+
+    uint32_t real_ntree_limit =
+        ntree_limit * model.learner_model_param->num_output_group;
+    if (real_ntree_limit == 0 || real_ntree_limit > model.trees.size()) {
+      real_ntree_limit = static_cast<uint32_t>(model.trees.size());
+    }
+
+    const int ngroup = model.learner_model_param->num_output_group;
+    CHECK_NE(ngroup, 0);
+    // allocate space for (number of features + bias) times the number of rows
+    std::vector<bst_float>& contribs = *out_contribs;
+    size_t contributions_columns =
+        model.learner_model_param->num_feature + 1;  // +1 for bias
+    contribs.resize(p_fmat->Info().num_row_ * contributions_columns *
+                    model.learner_model_param->num_output_group);
+    dh::caching_device_vector<float> phis(contribs.size(), 0.0);
+    // Add the bias term to last column
+    p_fmat->Info().base_margin_.SetDevice(generic_param_->gpu_id);
+    const auto margin = p_fmat->Info().base_margin_.ConstDeviceSpan();
+    CHECK(margin.empty());
+    float base_score = model.learner_model_param->base_score;
+    auto d_phis = phis.data().get();
+    dh::LaunchN(
+        generic_param_->gpu_id,
+        p_fmat->Info().num_row_ * model.learner_model_param->num_output_group,
+        [=] __device__(size_t idx) {
+          d_phis[(idx + 1) * contributions_columns - 1] =
+              margin.empty() ? base_score : margin[idx];
```

Comment:
Wouldn't you get out of bound error for forest or multi-class model?  Let's add a check before entering the kernel.

Reply:
Don't think so, what check do you want specifically? I added a test for margin.

Reply:
`CHECK_EQ(margin.Size(), p_fmat->Info().num_row_)` or `CHECK_EQ(margin.Size(), p_fmat->Info().num_row_ * num_groups)`?  I can't be sure.